java/monkey/src/main/java/Lexer.java

```diff
@@ -85,31 +73,33 @@ private char getCc() {
         return this.peek(this.pos);
     }
 
+    private char readCc() {
+        var currentChar = this.getCc();
+        this.advance();
+        return currentChar;
+    }
+
     private char peek(int pos) {
         if (pos >= this.input.length() || pos < 0) {
             return '\0';
         }
         return this.input.charAt(pos);
     }
 
-    private char peekNext() {
-        return peek(this.pos + 1);
-    }
-
     private boolean isLetter(char c) {
         return Character.isLetter(c) || c == '_';
     }
 
     private String number() {
-        int pos = this.pos;
+        int pos = this.pos - 1;
```

Comment:
Not the biggest fan of this change. I'd rather look forward than backwards. In my mind the Lexer is an `Iterator<Token>`, and so any data behind us is gone, and we can only glance at the next one.

Reply:
I think this makes more sense than what we were doing before, that is, selectively advancing the pos pointer depending on what token was being generated. That was needed because the identifier and number generation moved the pointer but the others didn't. Now we're following the book more closely: we consume the current char (read and move pos) at the beginning of nextToken, so we don't have to move it later. The `pos - 1` is only needed because the position pointer has already been moved once before starting to create the identifier or number string. Notice we are not moving the pointer (not going backwards) just creating a variable that represents where the string should being.